flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java

```diff
@@ -547,4 +548,43 @@ private void createManagedDirectory(Path managedPath) {
 
     @Override
     public void close() throws IOException {}
+
+    // ------------------------------------------------------------------------
+    //  restore
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void restoreStateHandles(
+            long checkpointId, SubtaskKey subtaskKey, Stream<SegmentFileStateHandle> stateHandles) {
+
+        Set<LogicalFile> uploadedLogicalFiles;
+        synchronized (lock) {
+            uploadedLogicalFiles =
+                    uploadedStates.computeIfAbsent(checkpointId, id -> new HashSet<>());
+        }
+
+        stateHandles.forEach(
+                fileHandle -> {
+                    // The file cleanup ownership of recovered physicalFile belongs to
+                    // jobManager, so we should set file deleter to "null" for them.
```

Comment:
IIUC,  JM is only aware of `SegmentFileStateHandle`, and the `SegmentFileStateHandle#discardState()` is empty(). Currently, the deletion of physicalFile relies on JM's checkpoint completed/abort notification, which is executed on the TM. So, the file deleter shouldn't be null.

Reply:
Well that depends. If this is a failover, the TM should take management of files, otherwise TM leave the files to JM.

Reply:
@fredia @Zakelly  Thanks for your feedback. I have modified it to reconstruct the restored PhysicalFile using "physicalFileDeleter" .